### PR TITLE
feat(SCM) Introduces a "local" SCM with autoguessing

### PR DESCRIPTION
**Pull request by dduportal**
Ref. https://github.com/updatecli/updatecli/issues/716

Depends on #717 

This change request introduces a new feature for Updatecli: an automatic SCM identified as `local`.

The "automatic guess" feature is not complete: for now, it only extracts some of the atrributes, the others have to be specified through manifest or env. vars. I expect more PRs to improve it along the way.
That's why the associated issue won't be closed.

For now:
- URL for git
- Owner/repository for github

Please note that:

- A new go module dependency was added: `github.com/goware/urlx`. Not strictly necessarly, but really useful to parse the hostname of the github SSH addresses. Might be doable with a good regexp, but that was too distracting for me to write such (as it needs tests, etc.).
- The `gitgeneric` package now provides an interface, implemented by the default `GoGit{}` struct. Should not change anything (plugins `git`, `github` and others have been updated) but it allows mocking (a basic mock is provided because used by the code introduced in this PR) and would help in unit testing any SCM-dependent behavior.
- The `config` package was split (`errors.go` to store the errors, `defaults.go` for the logic around setting up default configuration). Be careful with rebases!
- The `scm` package was split to have a clear source file separation between `scm.Scm` behaviors and `scm.Config` behaviors (Validate was confusing before)
- There are 2 `Merge` methods for `git` and `github` SCM plugins. This code could definitivelt be improved in the future, but it's around the whole "configuration definition" + "deep merge tree structures with zero-ed values". Too much compared to writing the 2 raws methods. Can be improved, but since it's tested and since we do not define new SCM plugins everyday, it's worth it for me.
- Unit tests, unit tests and unit tests
- The end to end manifests where modified to disable the `local` SCM as none of them use it (to avoid unplanned info/warning message).

## Test

To test this pull request, you can run the following commands:

```shell
make test
make test-e2e
```

## Additional Information

- The end to end tests where modified to make its easier to run locally.
  - A change has been introduced in the `Makefile` to explain the env. var `$VENOM_VAR_binpath` but also to fail fast with a comprehensive message when missing env vars are absent (instead of starting venom which fails)
  - We can still improve the contributor experience for E2E but later (splitting tests for easier "catch which onefailed") :)
  - No more `values.yaml` for now. If we want to end-to-end testing of the templating of values, we have to add a specific test for this. Otherwise it adds too much varying elements. Another idea: we can also make this file back, but the Makefile would parse it to list the required environment variables in a centralized location).
  -

### CODE REVIEW

File: pkg/core/config/main.go
Comment (olblak) [inline pkg/core/config/main.go:48]: This is just a suggestion

```suggestion
const (
    // LOCALSCMIDENTIFIER defines the scm id used to configure the local scm directory  
    LOCALSCMIDENTIFIER string = "local"
    localScmIdentifier = "local"
)
```

File: pkg/core/pipeline/scm/main.go
Comment (olblak) [inline pkg/core/pipeline/scm/main.go:22]:
```suggestion
        // Disabled is a setting used to disable the local git repository auto configuration
	Disabled bool
```

File: pkg/core/pipeline/scm/main.go
Comment (olblak) [inline pkg/core/pipeline/scm/main.go:131]: I am wondering if it's worthwhile to allow the configuration of originremote via the scm spec

File: pkg/core/pipeline/scm/config.go
Comment (olblak) [inline pkg/core/pipeline/scm/config.go:94]: What do you think to guess the username and password based on "GITHUB_ACTOR" and "GITHUB_TOKEN" to align with github action behavior

File: pkg/core/pipeline/scm/config.go
Comment (olblak) [inline pkg/core/pipeline/scm/config.go:95]: We should track that we need to automatically detech the branch as well in a separated PR/issue

File: pkg/core/pipeline/scm/config.go
Comment (dduportal) [reply pkg/core/pipeline/scm/config.go:94]: I would prefer to use the pattern proposed in the PR, with explicit UPDATECLI_ vars
- The UPDATECLI_ pattern is more deterministic and void accidental usages
- These 2 variables (`GITHUB_*`) are complicated to overrides (if you want to use another token with more permissions for instance)

File: e2e/updatecli.d/deprecated.d/transformers.yaml
Comment (lemeurherve) [inline e2e/updatecli.d/deprecated.d/transformers.yaml:5]: How about `enabled: false` instead? Seems more common to have a "positive" expression as field name

File: pkg/core/config/defaults.go
Comment (lemeurherve) [inline pkg/core/config/defaults.go:15]: The `[updatecli]` prefix is already harcoded like that elsewhere?

Comment (dduportal) [reply pkg/core/config/defaults.go:15]: I've only changed the method from one filte to the other, I won't touch this method on this PR.

But that is a good question that you raise though :)

File: e2e/updatecli.d/deprecated.d/transformers.yaml
Comment (dduportal) [reply e2e/updatecli.d/deprecated.d/transformers.yaml:5]: Depends on the languages 😅
- Human language: depends on the culture and people profiles and there isn't really a "best approach".
- Technical language (see below) with zero-ed value make it easier to have a "disabled" attribute

The idea is the following:

- Default behavior is "feature" is enabled
- You want to provide a switch to have disable it if required
- The "zero" value for a boolean in Golang (same problem as helm templates or Groovy language) is `false`. 
- If you want an "enable", then you have to write code for all initialization path that would check the inverted value, since it would be "enable:  false` by default. Much more code to write.

Comment (lemeurherve) [reply e2e/updatecli.d/deprecated.d/transformers.yaml:5]: You're lazy, I understand 😄 😉

File: pkg/core/pipeline/scm/config.go
Comment (olblak) [reply pkg/core/pipeline/scm/config.go:95]: Sounds good to me, we can have smaller PR later.

File: pkg/core/config/defaults.go
Comment (olblak) [reply pkg/core/config/defaults.go:15]: > The `[updatecli]` prefix is already harcoded like that elsewhere?

In the same function yes 😆 , that function could be better written but well 🤷🏾

File: pkg/core/pipeline/scm/config.go
Comment (olblak) [reply pkg/core/pipeline/scm/config.go:94]: That's a good point. I started using `UPDATECLI_GITHUB_TOKEN` so I don't conflict with Github Action environment but at the same time `GITHUB_TOKEN` and `GITHUB_ACTOR` became pretty standard so I am fine with both

**Comment by olblak** [comment]
Overall, it's a great PR. Thanks for paving the way 
Considering the size of the work needed here, I totally support merging this PR even if it doesn't solve the global issue.

But before merging I would like to clarify one thing

## Disable local scm 

I don't understand why and when to use the following block

```
scms:
  local:
    disabled: true
```

File: pkg/core/config/defaults.go
Comment (lemeurherve) [inline pkg/core/config/defaults.go:34]:
```suggestion
// EnsureLocalScm ensures the config receiver has a
// "local" SCM configuration if needed
```

File: pkg/core/config/errors.go
Comment (lemeurherve) [inline pkg/core/config/errors.go:18]:
```suggestion
	// ErrNoKeyDefined is returned when during the templating process, updatecli tries to
```

File: pkg/core/config/main.go
Comment (lemeurherve) [inline pkg/core/config/main.go:99]:
```suggestion
	// Ensure there is a local SCM defined as specified
```

File: pkg/core/pipeline/scm/config.go
Comment (lemeurherve) [inline pkg/core/pipeline/scm/config.go:75]:
```suggestion
		// Case of non HTTPS scheme: either git or SSH (github.com for instance).
		// Hostname extraction on the "left" of the colon character (':')
```
I don't see the point of using another type of comment.

File: pkg/core/pipeline/scm/main.go
Comment (lemeurherve) [inline pkg/core/pipeline/scm/main.go:23]:
```suggestion
// ScmHandler is an interface offering common functions for a source control manager like git or github
```

File: pkg/plugins/utils/gitgeneric/main_test.go
Comment (lemeurherve) [inline pkg/plugins/utils/gitgeneric/main_test.go:183]:
```suggestion
			// Because origin's URL, as well as other remotes, depends on the user or CI configuration and is not deterministic.
```

File: pkg/plugins/utils/gitgeneric/mocks.go
Comment (lemeurherve) [inline pkg/plugins/utils/gitgeneric/mocks.go:5]:
```suggestion
	GitHandler // Ensure any unspecified method from this interface are still declared
```

**Review by lemeurherve — DISMISSED** [review]
Minor comment nitpicks

Great job! 🎉

File: e2e/values.yaml
Comment (dduportal) [reply e2e/values.yaml:1]: Oh, I thought I removed it. good catch!

**Review by lemeurherve — APPROVED** [review]

**Comment by dduportal** [comment]
Rebase in progress (with conflict merge)

**Comment by dduportal** [comment]
@olblak Ready to review (approve/merge if ok).

I've done full testing on jenkins-infra/kubernetes-management as usual with a `updatecli diff` with this PR.
Also tested the full "apply" on https://github.com/dduportal/docker-hashicorp-tools/pulls (3 opened PRs).

**Review by olblak — APPROVED** [review]
🎉

**Pull request merged**
